include/cuda/memory_resource

```diff
@@ -0,0 +1,809 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+#include <memory>
+#include "std/cstddef"
+#include "std/detail/__config"
+#include "std/detail/__pragma_push"
+#include "std/type_traits"
+#include "std/utility"
+#include "std/version"
+#include "stream_view"
+
+#if _LIBCUDACXX_STD_VER > 14
+#include <memory_resource>
+#endif // _LIBCUDACXX_STD_VER > 14
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/*!
+ * \brief Groups the tag types denoting the kind of memory of an allocation.
+ *
+ * Memory allocation kind determines where memory can be accessed and the
+ * performance characteristics of accesses.
+ *
+ * This is not a closed set, the user code can define custom memory kinds.
+ */
+namespace memory_kind {
+  /*!
+   * \brief Ordinary host memory
+   */
+  struct host;
+
+  /*!
+   * \brief Device memory, as allocated by cudaMalloc.
+   */
+  struct device;
+
+  /*!
+   * \brief Device-accessible host memory.
+   */
+  struct pinned;
+
+  /*!
+   * \brief Virtual memory that is automatically migrated between the host and devices.
+   */
+  struct managed;
+};
+
+/*!
+ * \brief Tag type for the default context of `memory_resource`.
+ *
+ * Default context in which storage may be used immediately on any thread or any
+ * CUDA stream without synchronization.
+ */
+struct any_context{};
+
+namespace detail {
+template <typename _Context>
+class __get_context_impl {
+protected:
+  virtual _Context do_get_context() const noexcept = 0;
+};
+
+template<>
+class __get_context_impl<any_context> {
+protected:
+  virtual any_context do_get_context() const noexcept { return any_context{}; }
+};
+
+}  // namespace detail
+
+template <typename _ResourcePointer, typename... _Properties>
+class basic_resource_view;
+
+/*!
+ * \brief Groups the tag types denoting the execution environment in which the memory can be accessed
+ *
+ * This is not a closed set, the user code can define custom accessibility.
+ */
+namespace memory_access {
+    struct host;
+    struct device;
+}
+
+/*!
+ * \brief A memory property tag type indicating that the memory can be oversubscribed.
+ *
+ * Oversubscribable memory doesn't need to have backing physical storage at all times.
+ */
+struct oversubscribable;
+
+/*!
+ * \brief A memory property tag type indicating that the memory has a backing physical
+ *        storage in the target location at all times.
+ */
+struct resident;
```

Comment:
I don't like that these two things which are opposites do not have names indicating they are opposites.  Would the names "virtual_memory" and "physical_memory" work?

Reply:
Virtual memory and physical memory have a *very* strong association with the layers in memory management to me, and I think we should avoid that.

I really like "resident"; is there a word that hints at being the opposite of it somehow? Maybe "resident" and "pageable" or something like that would make a good pair? (Though admittedly the strict meaning of "pageable" isn't really what people associate with it the most, so it's probably also a problematic word to use here.)